core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -166,9 +167,16 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
                 if (oldPrimary == null) {
                     throw new IllegalStateException("missing primary shard for " + newShardTable.shardId());
                 }
-                if (newPrimary.assignedToNode() &&
-                        newPrimary.isSameAllocation(oldPrimary) == false &&
-                        (oldPrimary.relocating() && newPrimary.isSameAllocation(oldPrimary.buildTargetRelocatingShard())) == false) {
+                // we update the primary term on initial assignment or when a replica is promoted. Most notably we do *not*
+                // update them when a primary relocates
+                if (newPrimary.unassigned() ||
+                        newPrimary.isSameAllocation(oldPrimary) ||
+                        // we do not use newPrimary.isTargetRelocationOf(oldPrimary) because that one enforces newPrimary to
+                        // be initializing. However, when the target shard is activated, we still want the primary term to staty
+                        // the same
+                        (oldPrimary.relocating() && newPrimary.isSameAllocation(oldPrimary.buildTargetRelocatingShard()))) {
```

Comment:
I agree - this is confusing ( I got it wrong first too). I don't want to expand the scope of this PR too much , so prefer a follow up..